src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,144 @@
+#include <app/InteractionModelTimeout.h>
+#include <app/OperationalSessionSetup.h>
+#include <memory>
+
+namespace chip {
+namespace app {
+
+ICDRefreshKeyInfo::ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, ICDClientInfo * apICDClientInfo,
+                                     ICDClientStorage * aICDClientStorage) :
+    mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{
+    mpCheckInDelegate  = apCheckInDelegate;
+    mpICDClientInfo    = apICDClientInfo;
+    mpICDClientStorage = aICDClientStorage;
+}
+
+void RegisterCommandSenderDelegate::OnResponse(chip::app::CommandSender * apCommandSender,
+                                               const chip::app::ConcreteCommandPath & aPath, const chip::app::StatusIB & aStatus,
+                                               chip::TLV::TLVReader * aData)
+{
+    app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType responseData;
+    if (aData == nullptr)
+    {
+        mpICDRefreshKeyInfo->GetCheckInDelegate()->OnRegistrationUpdateFailure(mpICDRefreshKeyInfo.get(),
+                                                                               CHIP_ERROR_INVALID_TLV_ELEMENT);
+        Platform::Delete(this);
+        return;
+    }
+
+    mError = DataModel::Decode(*aData, responseData);
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to decode re-registration response data: %" CHIP_ERROR_FORMAT, mError.Format());
+        mpICDRefreshKeyInfo->GetCheckInDelegate()->OnRegistrationUpdateFailure(mpICDRefreshKeyInfo.get(), mError);
+        Platform::Delete(this);
+        return;
+    }
+    // Update the ICDClientInfo with new key and start counter and store it to persistence
+    mpICDRefreshKeyInfo->GetICDClientInfo()->start_icd_counter = responseData.ICDCounter;
+    mError = mpICDRefreshKeyInfo->GetICDClientStorage()->SetKey(*mpICDRefreshKeyInfo->GetICDClientInfo(),
+                                                                mpICDRefreshKeyInfo->mNewKey.Span());
+    if (mError != CHIP_NO_ERROR)
+    {
+        ChipLogError(ICD, "Failed to store the new key after re-registration: %" CHIP_ERROR_FORMAT, mError.Format());
+        Platform::Delete(this);
+    }
+    mpICDRefreshKeyInfo->GetCheckInDelegate()->OnRegistrationUpdateComplete(mpICDRefreshKeyInfo->GetICDClientInfo(),
+                                                                            mpICDRefreshKeyInfo.get());
+    Platform::Delete(this);
+}
+void RegisterCommandSenderDelegate::OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError)
+{
+    mError = aError;
+    mpICDRefreshKeyInfo->GetCheckInDelegate()->OnRegistrationUpdateFailure(mpICDRefreshKeyInfo.get(), mError);
+    Platform::Delete(this);
+}
+
+void RegisterCommandSenderDelegate::OnDone(chip::app::CommandSender * apCommandSender) {}
+
+void RegisterCommandSenderDelegate::OnFailure(CHIP_ERROR aError)
```

Comment:
OnFailure is dup for OnError, we just need one. thanks

Reply:
Removed OnFailure as it was not needed after redesign.